src/hash.cr

```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
     @block = block
   end
 
+  # Creates a new empty `Hash` of *initial_capacity* with a *block* that handles missing keys.
+  #
+  # The *initial_capacity* is useful to avoid unnecessary reallocations
+  # of the internal buffer in case of growth. If you have an estimate
+  # of the maximum number of elements a hash will hold, the hash should
+  # be initialized with that capacity for improved performance.
+  #
+  # NOTE: *initial_capacity* defaults to 11 and an input of < 11 is ignored.
+  # NOTE: `#size` does not reflect capacity as in `Array`.
+  #
+  # ```
+  # new_hash = Hash(String, Int32).new(50) do |hash, key|
+  #   key.size
+  # end
+  # new_hash.size # => 0
+  # new_hash["zero"] = 0
+  # new_hash["two"] = 2
```

Comment:
These two lines don't really help illustrate the purpose. I think this would be a better example:
```cr
hash = Hash(String, Int32).new do |hash, key|
  hash[key] = key.size
end

hash.size        # => 0
hash["foo"]      # => 3
hash.size        # => 1
hash["bar"] = 10
hash["bar"]      # => 10
```

Reply:
Sounds good to me. I was also thinking of borrowing the "go fish" example used in the Ruby docs. Does anyone want to comment on the value of *that* example as opposed to this one?

They also had one where you used the block feature to essentially use the hash as a cache for some kind of calculation that can be done using the key alone. But I wondered if that was too abstract.